No newline at end of file

hi! thanks for this project.

after `uninstall`ing a couple dependencies, `git diff` reports my changes which include

<img width="262" alt="Screen Shot 2020-04-28 at 12 19 40 PM" src="https://user-images.githubusercontent.com/820696/80511806-99d9e280-894a-11ea-9eb3-6e0c53b7de0d.png">

it seems like `elm` produces an `elm.json` which includes a trailing newline, but this project does not include the newline. maybe i'll dig in and send you a PR

thanks @harrysarson! i tried that route, had trouble, and then thought about this `Serialize` route.

maybe i'm misunderstand, but putting `writeln!(&mut writer, "");` on line 63...

https://github.com/zwilias/elm-json/blob/79ac385c4e13e7f2fd3cb9fc84f6723a93275718/src/lib/cli/util.rs#L54-L64

...results in this error

```
error[E0596]: cannot borrow `writer` as mutable, as it is not declared as mutable
  --> src/lib/cli/util.rs:63:14
   |
57 |     let writer = BufWriter::new(file);
   |         ------ help: consider changing this to be mutable: `mut writer`
...
63 |     writeln!(&mut writer, "");
   |              ^^^^^^^^^^^ cannot borrow as mutable

error[E0382]: borrow of moved value: `writer`
  --> src/lib/cli/util.rs:63:14
   |
57 |     let writer = BufWriter::new(file);
   |         ------ move occurs because `writer` has type `std::io::BufWriter<std::fs::File>`, which does not implement the `Copy` trait
58 |     let formatter = serde_json::ser::PrettyFormatter::with_indent(b"    ");
59 |     let mut serializer = serde_json::Serializer::with_formatter(writer, formatter);
   |                                                                 ------ value moved here
...
63 |     writeln!(&mut writer, "");
   |              ^^^^^^^^^^^ value borrowed here after move

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0382, E0596.
For more information about an error, try `rustc --explain E0382`.
error: could not compile `elm-json`.
```

i _kind of_ understand ownership but not well enough to see the solution here. my understanding is that the ownership was passed to the serializer, which led me to thinking i should modify how things get serialized.

--- a/src/lib/cli/util.rs
+++ b/src/lib/cli/util.rs
@@ -9,12 +9,7 @@ use dialoguer::Confirmation;
 use failure::{format_err, ResultExt};
 use serde::ser::Serialize;
 use slog::Logger;
-use std::{
-    collections::HashSet,
-    convert,
-    fs::File,
-    io::{BufReader, BufWriter},
-};
+use std::{collections::HashSet, convert, fs::File, io::Write};
 
 pub fn confirm(prompt: &str, matches: &ArgMatches) -> Result<bool> {
     if matches.is_present("yes") {
@@ -46,20 +41,19 @@ where
 fn read_elm_json(matches: &ArgMatches) -> Result<Project> {
     let path = matches.value_of("INPUT").unwrap();
     let file = File::open(path).context(ErrorKind::MissingElmJson)?;
-    let reader = BufReader::new(file);
-    let info: Project = serde_json::from_reader(reader).context(ErrorKind::InvalidElmJson)?;
+    let info: Project = serde_json::from_reader(file).context(ErrorKind::InvalidElmJson)?;
     Ok(info)
 }
 
 pub fn write_elm_json(project: &Project, matches: &ArgMatches) -> Result<()> {
     let path = matches.value_of("INPUT").unwrap();
-    let file = File::create(path).context(ErrorKind::UnwritableElmJson)?;
-    let writer = BufWriter::new(file);
+    let mut file = File::create(path).context(ErrorKind::UnwritableElmJson)?;
     let formatter = serde_json::ser::PrettyFormatter::with_indent(b"    ");
-    let mut serializer = serde_json::Serializer::with_formatter(writer, formatter);
+    let mut serializer = serde_json::Serializer::with_formatter(&file, formatter);
     project
         .serialize(&mut serializer)
         .context(ErrorKind::Unknown)?;
+    write!(file, "\n").context(ErrorKind::Unknown)?;
     Ok(())
 }
 

